src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -437,33 +454,131 @@ public static unsafe void AddScalarU(float scalar, Span<float> dst)
+                    }
+                    return;
+                }
 
-                    dstVector = Sse.Multiply(scaleVector, dstVector);
-                    Sse.Store(pDstCurrent, dstVector);
+                nuint address = (nuint)(pd);
+                int misalignment = (int)(address % 16);
+                int remainder = 0;
 
-                    pDstCurrent += 4;
-                }
+                if ((misalignment & 3) != 0)
+                {
+                    // Handles cases where the data is not 32-bit aligned and we can't ever use aligned operations
+                    remainder = length % 4;
 
-                while (pDstCurrent < pEnd)
+                    for (float* pEnd = pd + (length - remainder); pDstCurrent < pEnd; pDstCurrent += 4)
+                    {
+                        Vector128<float> temp = Sse.LoadVector128(pDstCurrent);
+                        temp = Sse.Multiply(scaleVector128, temp);
+                        Sse.Store(pDstCurrent, temp);
+                    }
+                }
+                else
                 {
-                    Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+                    if (misalignment != 0)
+                    {
+                        // Handle cases where the data is not 128-bit aligned by doing an unaligned read and then
+                        // masking any elements that will be included in the first aligned read
 
-                    dstVector = Sse.MultiplyScalar(scaleVector, dstVector);
-                    Sse.StoreScalar(pDstCurrent, dstVector);
+                        misalignment >>= 2;
+                        misalignment = 4 - misalignment;
 
-                    pDstCurrent++;
+                        Vector128<float> result = Sse.LoadVector128(pDstCurrent);
+
+                        Vector128<float> leadingMask = Sse.LoadVector128(((float*)(pLeadingAlignmentMask)) + (misalignment * 4));
+                        Vector128<float> trailingMask = Sse.LoadVector128(((float*)(pTrailingAlignmentMask)) + ((4 - misalignment) * 4));
+
+                        Vector128<float> temp = Sse.And(result, leadingMask);
+                        result = Sse.And(result, trailingMask);
+
+                        temp = Sse.Multiply(scaleVector128, temp);
+                        result = Sse.Or(temp, result);
+
+                        Sse.Store(pDstCurrent, result);
+
+                        pDstCurrent += misalignment;
+                        length -= misalignment;
+                    }
+
+                    if (length > 4)
+                    {
+                        // Handle all the 128-bit blocks that we can now that we have offset to an aligned address
+
+                        remainder = length % 4;
+
+                        for (float* pEnd = pDstCurrent + (length - remainder); pDstCurrent < pEnd; pDstCurrent += 4)
+                        {
+                            if (Avx.IsSupported)
```

Comment:
This is dead code, isn't it?  if Avx.IsSupported, we would never get into the `SseIntrinsics` class.

Reply:
Not quite dead as we cover it in tests today.

We should also be setting the AVX path to only execute if supports and if the input data is over a certain size (given that it can cause temporary downclocking for that core and impact neighboring workloads).

Reply:
The only place in the product this method is called has:

https://github.com/dotnet/machinelearning/blob/1427f621c42c1da576c653bf399f38c66d69c00f/src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs#L247-L257

Having this in here just for our perf tests doesn't make sense because no actual user is going to be able to get into this code.

Reply:
Yes, but it is something that we should be allowing to execute before shipping.

As per the Intel Architecture Optimization Manual, executing even a single 256-bit (or higher) AVX instruction can cause the core to temporarily downclock to a lower frequency. For small workloads, this can cause subsequent non 256-bit workloads to execute more slowly until the processor clocks up again.

This means that we need to determine what a "good" threshold is for each of these and only allow the 256-bit code path to execute `if (Avx.IsSupported && (dst.Length >= threshold))`

Reply:
so whats the final verdict on this ?

Reply:
@tannergooding - can you log an issue for your thoughts here?

I would prefer that (for now) we don't introduce dead code into the product. We can add it later when deemed necessary. If I was coming into this codebase, and saw that this check existed inside a method that is guaranteed never to be called in that case, I would remove the dead code.

@justinormont - I think you are meaning to comment on #1028.

Reply:
what the final verdict on this ? should I go ahead and remove that block ?

Reply:
cc @eerhardt @tannergooding

Reply:
I have removed this path for now